### PR TITLE
HADOOP-18896. Large `file.bytes-per-checksum` cause NegativeArraySizeException in FSOutputSummer

### PR DESCRIPTION
**Pull request by anshuksi282-ksolves**
## Description of PR
This PR fixes HADOOP-18896.

The buffer size in `FSOutputSummer` is computed as `sum.getBytesPerChecksum() * BUFFER_NUM_CHUNKS`. When
`file.bytes-per-checksum` is configured to a very large value (e.g. 238609295), this multiplication overflows a signed 32-bit int and wraps around to a negative number, causing a `NegativeArraySizeException` when allocating the internal buffer array.

This PR adds a `Preconditions.checkArgument` check in the `FSOutputSummer` constructor to validate that the computed buffer size is positive, failing fast with a clear `IllegalArgumentException` instead of an obscure `NegativeArraySizeException`.

This continues the work started in #6064 by @teamconfx, which went stale. This PR additionally adds a unit test, as the original PR's `test4tests` check had failed for not including one.

### How was this patch tested?
- Added a new test class `TestFSOutputSummer` in `hadoop-common` with `testLargeBytesPerChecksumOverflow`, using     a minimal concrete subclass of `FSOutputSummer` to verify the constructor fails with `IllegalArgumentException` (rather than `NegativeArraySizeException`) when `bytesPerChecksum * BUFFER_NUM_CHUNKS` overflows.
- Verified the test fails with the original `NegativeArraySizeException` when the fix is reverted, and passes with the fix applied.
- `mvn -pl hadoop-common-project/hadoop-common test -Dtest=TestFSOutputSummer` passes.

Note: the JIRA's original reproduction steps reference an HDFS test (`TestDecommissionWithStriped`), but since the buffer overflow bug lives entirely in `hadoop-common`'s `FSOutputSummer` class, this PR adds a focused unit test in `hadoop-common` instead, avoiding an unnecessary cross-module HDFS test dependency.

### AI Tooling
Contains content generated by Claude (Anthropic). Used to help investigate the root cause, draft the fix, and write/verify the accompanying unit test.

### CODE REVIEW

**Comment by anshuksi282-ksolves** [comment]
Hi all,
Requesting a review whenever you get a chance. This PR picks up HADOOP-18896 from where #6064 (@teamconfx) left off — it went stale without a unit test. I've added a unit test (`TestFSOutputSummer#testLargeBytesPerChecksumOverflow`) that reproduces the original `NegativeArraySizeException` without the fix, and verifies the fix resolves it with a clean `IllegalArgumentException` instead.

I'm new to contributing to Hadoop, so please let me know if any changes are needed. Thanks in advance!

cc: @ayushtkn @dineshchitlangia @brahmareddybattula @pjfanning @slfan1989

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FSOutputSummer.java [hunk 1]
```diff
@@ -52,6 +53,9 @@ abstract public class FSOutputSummer extends OutputStream implements
   
   protected FSOutputSummer(DataChecksum sum) {
     this.sum = sum;
+    Preconditions.checkArgument(
+            sum.getBytesPerChecksum() * BUFFER_NUM_CHUNKS > 0,
```
Comment (pjfanning) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FSOutputSummer.java]: I would do something like:
```
int bufArraySize;
try {
  bufArraySize = Math.multiplyExact(sum.getBytesPerChecksum(), BUFFER_NUM_CHUNKS);
} catch (ArithmeticException ae) {
  throw new IllegalArgumentException("The calculated buffer array size for FSOutputSummer is too large", ae);
}
if (bufArraySize < 0) {
  throw new IllegalArgumentException("The calculated buffer array size for FSOutputSummer is negative");
}
this.buf = new byte[bufArraySize];
```
I would suggest that you also check the array size in a similar way for `this.checksum`.

Comment (anshuksi282-ksolves) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FSOutputSummer.java]: Thanks for the review @pjfanning! 
Updated to use `Math.multiplyExact` for both `buf` and `checksum` buffer size checks, as suggested.

I did try adding a dedicated test for the `checksum` buffer overflow path, but `getChecksumSize()` returns a fixed value from `DataChecksum.Type` (e.g. 4 for CRC32), so it can never actually be large enough to overflow — that path isn't reachable in practice, just defensive. So I've kept the existing unit test covering the `buf` overflow case, and re-ran `TestFSOutputSummer`, `TestChecksumFileSystem`, and `TestChecksumFs` to confirm no regressions.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   9m  8s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  27m  1s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   8m 40s |  |  trunk passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  compile  |   8m 55s |  |  trunk passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  checkstyle  |   0m 49s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 11s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 55s |  |  trunk passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javadoc  |   0m 49s |  |  trunk passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  spotbugs  |   1m 45s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  18m 11s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 40s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   8m 31s |  |  the patch passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javac  |   8m 31s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   9m 45s |  |  the patch passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javac  |   9m 45s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 47s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 17s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 46s |  |  the patch passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javadoc  |   0m 43s |  |  the patch passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  spotbugs  |   1m 57s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  18m 13s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 20s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 40s |  |  The patch does not generate ASF License warnings.  |
|  |   | 139m 56s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.55 ServerAPI=1.55 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-8622/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/8622 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 5ab2ba19cd5b 5.15.0-181-generic #191-Ubuntu SMP Fri May 22 19:09:02 UTC 2026 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 5985a438245d40c27c554e7d00f7366ba9650cb5 |
| Default Java | Ubuntu-17.0.19+10-1-24.04.2-Ubuntu |
| Multi-JDK versions | /usr/lib/jvm/java-21-openjdk-amd64:Ubuntu-21.0.11+10-1-24.04.2-Ubuntu /usr/lib/jvm/java-17-openjdk-amd64:Ubuntu-17.0.19+10-1-24.04.2-Ubuntu |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-8622/1/testReport/ |
| Max. process+thread count | 1264 (vs. ulimit of 10000) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-8622/1/console |
| versions | git=2.43.0 maven=3.9.15 spotbugs=4.9.7 |
| Powered by | Apache Yetus 0.14.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 24s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  28m 42s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   9m 41s |  |  trunk passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  compile  |   9m 53s |  |  trunk passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  checkstyle  |   0m 52s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m  5s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 47s |  |  trunk passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javadoc  |   0m 47s |  |  trunk passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  spotbugs  |   1m 40s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  18m 38s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 44s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   9m 48s |  |  the patch passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javac  |   9m 48s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   8m 49s |  |  the patch passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javac  |   8m 49s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 49s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m  9s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 54s |  |  the patch passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javadoc  |   0m 52s |  |  the patch passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  spotbugs  |   1m 53s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  16m 45s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 44s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 48s |  |  The patch does not generate ASF License warnings.  |
|  |   | 135m 26s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.55 ServerAPI=1.55 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-8622/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/8622 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 13c8d3582b82 5.15.0-181-generic #191-Ubuntu SMP Fri May 22 19:09:02 UTC 2026 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 60af129478b0beaaf388b80cd4af3572e43cf1a7 |
| Default Java | Ubuntu-17.0.19+10-1-24.04.2-Ubuntu |
| Multi-JDK versions | /usr/lib/jvm/java-21-openjdk-amd64:Ubuntu-21.0.11+10-1-24.04.2-Ubuntu /usr/lib/jvm/java-17-openjdk-amd64:Ubuntu-17.0.19+10-1-24.04.2-Ubuntu |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-8622/2/testReport/ |
| Max. process+thread count | 3155 (vs. ulimit of 10000) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-8622/2/console |
| versions | git=2.43.0 maven=3.9.15 spotbugs=4.9.7 |
| Powered by | Apache Yetus 0.14.1 https://yetus.apache.org |


This message was automatically generated.

**Review by dineshchitlangia — APPROVED** [review]
+1 LGTM

**Comment by anshuksi282-ksolves** [comment]
> +1 LGTM

Hi @dineshchitlangia @pjfanning, thanks for the reviews and the approval!

Since the CI is completely green and the requested changes have been addressed, could you please help get this merged? Let me know if there is anything else required from my end.

**Comment by pan3793** [comment]
Reviewed `FSOutputSummer.java` together with the `DataChecksum` factory and checksum sizes. The fix correctly turns the overflow into a fast `IllegalArgumentException`, and the test is well-targeted (`238609295 * 9 = 2147483655 > Integer.MAX_VALUE`). Nothing in the repo catches `NegativeArraySizeException`, so the exception-type change is safe. Findings below are simplifications, not correctness issues.

**1. Dead `if (bufSize < 0)` / `if (checksumBufSize < 0)` branches (minor)**

`Math.multiplyExact` throws `ArithmeticException` on overflow and never returns a wrapped negative, and `DataChecksum.newDataChecksum` already rejects `bytesPerChecksum <= 0` (`DataChecksum.java:112`), so `bufSize` is always positive after a successful multiply. Both negative checks are unreachable and can be dropped. (The PR description mentions a single `Preconditions.checkArgument`, but the implementation uses `Math.multiplyExact` + try/catch + the dead checks.)

**2. Checksum-buffer try/catch is dead (nit)**

`getChecksumSize()` returns `type.size`, which is `0` (NULL) or `4` (CRC32/CRC32C) only -- `4 * 9 = 36` cannot overflow. The original one-liner `new byte[getChecksumSize() * BUFFER_NUM_CHUNKS]` was already safe.

**3. `resetChecksumBufSize()` still uses raw multiplication (latent, minor)**

`FSOutputSummer#resetChecksumBufSize` computes `sum.getBytesPerChecksum() * BUFFER_NUM_CHUNKS` without the overflow guard. Not reachable with an overflowing value today (the constructor now rejects them and `sum` is final), but the two paths for the same product are now inconsistent. Consider routing both through one `Math.multiplyExact` helper.

**4. Exception type consistency (informational)**

`DFSOutputStream`'s own validation throws `HadoopIllegalArgumentException`; the new plain `IllegalArgumentException` differs. Just noting the inconsistency, not a required change.

**Comment by anshuksi282-ksolves** [comment]
Hi @pan3793,

Thank you for the detailed and insightful review! I have pushed a new commit addressing all your findings:

1. Removed the dead if (bufSize < 0) and if (checksumBufSize < 0) branches, as Math.multiplyExact handles the overflow by throwing an ArithmeticException.
2. Reverted the checksum buffer initialization back to the simple new byte[getChecksumSize() * BUFFER_NUM_CHUNKS] one-liner, since getChecksumSize() will never be large enough to overflow.
3. Updated resetChecksumBufSize() to use Math.multiplyExact for consistency with the constructor and to guard against any latent overflow issues.
4. Regarding the exception type, I've left it as IllegalArgumentException for now as you noted it's just informational, but please let me know if you'd prefer me to switch it to HadoopIllegalArgumentException.

Please let me know if everything looks good now!

**Review by pan3793 — APPROVED** [review]

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 24s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  26m 54s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   8m 38s |  |  trunk passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  compile  |   8m 51s |  |  trunk passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  checkstyle  |   0m 50s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 13s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 55s |  |  trunk passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javadoc  |   0m 52s |  |  trunk passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  spotbugs  |   1m 44s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 47s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 43s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   8m 17s |  |  the patch passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javac  |   8m 17s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   8m 50s |  |  the patch passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javac  |   8m 50s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 53s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 11s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 54s |  |  the patch passed with JDK Ubuntu-21.0.11+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  javadoc  |   0m 50s |  |  the patch passed with JDK Ubuntu-17.0.19+10-1-24.04.2-Ubuntu  |
| +1 :green_heart: |  spotbugs  |   1m 54s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  17m  5s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 42s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 47s |  |  The patch does not generate ASF License warnings.  |
|  |   | 128m 19s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.55 ServerAPI=1.55 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-8622/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/8622 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux f196ace275ed 5.15.0-181-generic #191-Ubuntu SMP Fri May 22 19:09:02 UTC 2026 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 204056eddb7de81edd6c1f1c607433bf482e3f31 |
| Default Java | Ubuntu-17.0.19+10-1-24.04.2-Ubuntu |
| Multi-JDK versions | /usr/lib/jvm/java-21-openjdk-amd64:Ubuntu-21.0.11+10-1-24.04.2-Ubuntu /usr/lib/jvm/java-17-openjdk-amd64:Ubuntu-17.0.19+10-1-24.04.2-Ubuntu |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-8622/3/testReport/ |
| Max. process+thread count | 3154 (vs. ulimit of 10000) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-8622/3/console |
| versions | git=2.43.0 maven=3.9.15 spotbugs=4.9.7 |
| Powered by | Apache Yetus 0.14.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by anshuksi282-ksolves** [comment]
Hi @pan3793 and @dineshchitlangia, I have seen you both have approved this PR but till now no one have mergerd it.
Is there anything needed from my end ?
